### PR TITLE
Add unit test for PipelineTableMetaData

### LINKED ISSUE #15698: [OpenForce 2022] Add unit test for PipelineTableMetaData
Hi community,

This issue is for [OpenForce 2022](https://github.com/apache/shardingsphere/discussions/12683#discussioncomment-2139441)

### Aim
Add unit test for `PipelineTableMetaData` in `shardingsphere-pipeline-test` module.

### Basic Qualifications
- Java
- Maven
- Junit.Test

### Example FYI
- `TableMetaDataTest`

### PR DESCRIPTION
**Pull request by pk-218**
Fixes #15698.

### CODE REVIEW

File: shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
+package org.apache.shardingsphere.data.pipeline.core.metadata.model;
+
```
Comment (sandynz) [inline shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java:2]: License header should be added at the beginning of java code. You could refer to other source code.

And it's better to run unit test before submiting PR. How to do it?
1. In terminal, cd to `shardingsphere-test/shardingsphere-pipeline-test` directory (assume it's in shardingsphere project root)
2. Execute command: `mvn clean install -Dmaven.javadoc.skip=true -Prelease`, fix error and run it again until it pass.

File: shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+import java.sql.Types;
+import java.util.List;
+import java.util.Map;
+
+public final class PipelineTableMetaDataTest {
+    private PipelineTableMetaData pipelineTableMetaData;
```
Comment (sandynz) [inline shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java:16]: From code of conduct: `Keep one blank line after class definition.`

See [Code of Conduct]( https://shardingsphere.apache.org/community/en/contribute/code-conduct/ ) for more details.

File: shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Test
+    public void assertGetColumnMetaDataGivenColumnName() {
+        PipelineColumnMetaData actual = columnMetaDataMap.get("test_column");
+        assertThat(actual.getOrdinalPosition(), is(0));
+        assertThat(actual.getName(), is("test"));
+        assertThat(actual.getDataType(), is(Types.INTEGER));
+        assertTrue(actual.isPrimaryKey());
+    }
```
Comment (sandynz) [inline shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java:43]: 1. It's better to invoke `pipelineTableMetaData.getColumnMetaData` instead of `columnMetaDataMap.get("test_column")`, since the logic is encapsulated in `pipelineTableMetaData` and we'll test it.

2. It could be better to add a new case: get column metadata by non-existence column name.

File: shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        PipelineColumnMetaData pipelineColumnMetaData = new PipelineColumnMetaData(0, "test", Types.INTEGER, "INTEGER", true);
+        columnMetaDataMap.put("test_column", pipelineColumnMetaData);
+        pipelineTableMetaData = new PipelineTableMetaData(null, columnMetaDataMap);
+    }
+
+    @Test
+    public void assertGetColumnMetaDataGivenColumnIndex() {
+        List<String> pipelineTableColumns = pipelineTableMetaData.getColumnNames();
+        PipelineColumnMetaData actual = columnMetaDataMap.get(pipelineTableColumns.get(0));
+        assertThat(actual.getOrdinalPosition(), is(0));
```
Comment (sandynz) [inline shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java:30]: 1. It's better to invoke `pipelineTableMetaData.getColumnMetaData` instead of `columnMetaDataMap.get(pipelineTableColumns.get(0))`, since the logic is encapsulated in `pipelineTableMetaData` and we'll test it.

2. `columnMetaDataMap` could be hidden in `setUp`.

**Review by sandynz — CHANGES_REQUESTED** [review]
Code is clean, that's good!

There's some little places could be improved.

File: shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java (hunk 3 above)
Comment (pk-218) [reply shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java:43]: 1. Thank you for pointing it out, I was a bit skeptical in using `columnMetaDataMap` for accessing the `test_column`. I have made the change.
2. I have added the new case -
```java
assertThat(pipelineTableMetaData.getColumnMetaData("not_exist"), is(-1));
```

File: shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java (hunk 4 above)
Comment (pk-218) [reply shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java:30]: 1. Yes agreed, I have made the change.
2. Shifted `columnMetaDataMap` declaration to `setUp`

File: shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java (hunk 2 above)
Comment (pk-218) [reply shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java:16]: Done

File: shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java (hunk 1 above)
Comment (pk-218) [reply shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java:2]: Added the license header.

Earlier I was trying to run ```./mvnw clean install `-Dmaven`.`javadoc`.skip=true -Prelease``` at project root since I am running the project on Windows. I ran into several errors with several tests of the  `shardingsphere-sql-parser-engine`  module failing. 

I will try running the unit tests again in the `shardingsphere-pipeline-test` directory.

File: shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java (hunk 4 above)
Comment (pk-218) [reply shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java:30]: I have done away with `columnMetaDataMap`. This is how `setUp` method looks now

```java
    @Before
    public void setUp() {
        pipelineTableMetaData = new PipelineTableMetaData(null, Collections.singletonMap("test_column", new PipelineColumnMetaData(0, "test", Types.INTEGER, "INTEGER", true)));
    }
```

Comment (sandynz) [reply shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java:30]: Nice. You could commit the change and push again, I'll enable the CI workflow then.

File: shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java (hunk 3 above)
Comment (pk-218) [reply shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java:43]: @sandynz I get a failing test case here while checking for a non-existent column since `getColumnMetaData` returns `null` here, not `-1`.

Comment (sandynz) [reply shardingsphere-test/shardingsphere-pipeline-test/src/test/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaDataTest.java:43]: You could try `assertNull(pipelineTableMetaData.getColumnMetaData("not_exist"))`.

**Comment by pk-218** [comment]
@sandynz I ran the tests again and I have 2 errors, both occur while passing `columnIndex = 0`
```java
PipelineColumnMetaData actual = pipelineTableMetaData.getColumnMetaData(0);
assertTrue(pipelineTableMetaData.isPrimaryKey(0));
```

This is the log obtained on running the tests.
```
[ERROR] Tests run: 3, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.002 s <<< FAILURE! - in org.apache.shardingsphere.data.pipeline.core.metadata.model
.PipelineTableMetaDataTest
[ERROR] assertGetColumnMetaDataGivenColumnIndex(org.apache.shardingsphere.data.pipeline.core.metadata.model.PipelineTableMetaDataTest)  Time elapsed: 0.002 s
  <<< ERROR!
java.lang.NullPointerException
        at org.apache.shardingsphere.data.pipeline.core.metadata.model.PipelineTableMetaDataTest.assertGetColumnMetaDataGivenColumnIndex(PipelineTableMetaDat
aTest.java:44)

[ERROR] assertIsPrimaryKey(org.apache.shardingsphere.data.pipeline.core.metadata.model.PipelineTableMetaDataTest)  Time elapsed: 0 s  <<< ERROR!
java.lang.NullPointerException
        at org.apache.shardingsphere.data.pipeline.core.metadata.model.PipelineTableMetaDataTest.assertIsPrimaryKey(PipelineTableMetaDataTest.java:59)
```


Could you please review why these errors are occurring? Or should I first push the code with new changes for review?

**Comment by sandynz** [comment]
Hi @pk-218 , you could push the code with new changes, I could not see the latest code to check why NPE occur for now.
I'm afraid to review the changes at tomorrow, since it's late-night here.

**Comment by pk-218** [comment]
I understand. I will push the changes for now. Thanks for taking out the time!

**Comment by sandynz** [comment]
Hi @pk-218 , I'm sorry, there's no javadoc to explain `columnMetaDataMap` (we could improve it).

From source code, `PipelineTableMetaData` is initialized in `PipelineTableMetaDataLoader`, related code snippet:
```
                boolean primaryKey = primaryKeys.contains(columnName);
                PipelineColumnMetaData columnMetaData = new PipelineColumnMetaData(ordinalPosition, columnName, dataType, dataTypeName, primaryKey);
                columnMetaDataMap.put(columnName, columnMetaData);
```
`columnMetaDataMap`'s key is column name, and also `PipelineColumnMetaData.name`.

In our test code `Collections.singletonMap("test_column", new PipelineColumnMetaData(1, "test", Types.INTEGER, "INTEGER", true))`. In short, `test_column` doesn't match `test`, that's why there's NPE.

And also, we could debug the failed test case in IDE to check every step, verify why variable is null that cause NPE.
A variables screenshot on my local debug:
图片

**Comment by pk-218** [comment]
Thank you for clearing it! All test cases pass now

**Comment by sandynz** [comment]
> Thank you for clearing it! All test cases pass now

Nice!
CI workflow is started, I'll merge it once it pass.

**Review by sandynz — APPROVED** [review]

**Comment by sandynz** [comment]
Merged.
@pk-218 , thanks for your contribution.
